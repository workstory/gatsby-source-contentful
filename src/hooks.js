// @ts-check
import { getImageData } from "gatsby-plugin-image"
import { useMemo } from "react"

import { createUrl } from "./image-helpers"

export function useContentfulImage({ image, ...props }) {
  return useMemo(
    () =>
      getImageData({
        baseUrl: image.url,
        sourceWidth: image.width,
        sourceHeight: image.height,
        backgroundColor: null,
        urlBuilder: ({ baseUrl, width, height, options, format }) =>
          createUrl(baseUrl, { ...options, height, width, toFormat: format }),
        pluginName: `gatsby-source-contentful`,
        ...props,
      }),
    [image.url, image.width, image.height, props]
  )
}
