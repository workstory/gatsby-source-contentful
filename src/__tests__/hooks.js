import { useContentfulImage } from "../hooks"

jest.mock(`react`, () => {
  return {
    ...jest.requireActual(`react`),
    useMemo: jest.fn(v => v()),
  }
})

describe(`useContentfulImage`, () => {
  const consoleWarnSpy = jest.spyOn(console, `warn`)

  const image = { url: `//images.ctfassets.net/foo/bar/baz/image.jpg` }

  beforeEach(() => {
    consoleWarnSpy.mockClear()
  })

  test(`constrained: width and aspectRatio`, () => {
    const result = useContentfulImage({
      image,
      width: 600,
      aspectRatio: 1.778,
    })
    const desiredHeight = Math.floor(600 / 1.778)
    expect(result.layout).toBe(`constrained`)
    expect(result.height).toBe(desiredHeight)
    expect(result.width).toBe(600)
    expect(result.images.fallback.src).toContain(`h=${desiredHeight}`)
    expect(result).toMatchSnapshot()
    expect(consoleWarnSpy).toHaveBeenCalledTimes(0)
  })

  test(`constrained: width and height`, () => {
    const result = useContentfulImage({
      image,
      width: 600,
      height: 480,
    })
    expect(result.layout).toBe(`constrained`)
    expect(result.height).toBe(480)
    expect(result.width).toBe(600)
    expect(result.images.fallback.src).toContain(`h=480`)
    expect(result.images.fallback.src).toContain(`w=600`)
    expect(result.images.fallback.srcSet.split(`,`)).toHaveLength(3)
    expect(result).toMatchSnapshot()
    expect(consoleWarnSpy).toHaveBeenCalledTimes(0)
  })

  test(`fixed: width and height`, () => {
    const result = useContentfulImage({
      layout: `fixed`,
      image,
      width: 600,
      height: 480,
    })
    expect(result.layout).toBe(`fixed`)
    expect(result.height).toBe(480)
    expect(result.width).toBe(600)
    expect(result.images.fallback.sizes).toContain(`600px`)
    expect(result.images.fallback.src).toContain(`h=480`)
    expect(result.images.fallback.src).toContain(`w=600`)
    expect(result.images.fallback.srcSet.split(`,`)).toHaveLength(1)
    expect(result).toMatchSnapshot()
    expect(consoleWarnSpy).toHaveBeenCalledTimes(0)
  })

  test(`fullWidth: aspect ratio`, () => {
    const result = useContentfulImage({
      image,
      layout: `fullWidth`,
      aspectRatio: 1.778,
    })
    expect(result.layout).toBe(`fullWidth`)
    expect(result.height).toBe(0.5625)
    expect(result.width).toBe(1)
    expect(result.images.fallback.sizes).toBe(`100vw`)
    expect(result.images.fallback.src).toContain(`h=2304`)
    expect(result.images.fallback.src).toContain(`w=4096`)
    expect(result.images.fallback.srcSet.split(`,`)).toHaveLength(12)
    expect(result).toMatchSnapshot()
  })

  test(`fullWidth: aspectRatio, maxWidth`, () => {
    const result = useContentfulImage({
      image,
      layout: `fullWidth`,
      aspectRatio: 1.778,
      maxWidth: 1280,
    })
    expect(result.layout).toBe(`fullWidth`)
    expect(result.height).toBe(0.5625)
    expect(result.width).toBe(1)
    expect(result.images.fallback.sizes).toBe(`100vw`)
    expect(result.images.fallback.src).toContain(`h=2304`)
    expect(result.images.fallback.src).toContain(`w=4096`)
    expect(result.images.fallback.srcSet.split(`,`)).toHaveLength(12)
    expect(result).toMatchSnapshot()
  })
  test(`fullWidth: width, height & maxWidth`, () => {
    const result = useContentfulImage({
      image,
      layout: `fullWidth`,
      width: 800,
      height: 600,
      maxWidth: 1280,
    })
    expect(result.layout).toBe(`fullWidth`)
    expect(result.height).toBe(0.75)
    expect(result.width).toBe(1)
    expect(result.images.fallback.sizes).toBe(`100vw`)
    expect(result.images.fallback.src).toContain(`h=600`)
    expect(result.images.fallback.src).toContain(`w=800`)
    expect(result.images.fallback.srcSet.split(`,`)).toHaveLength(4)
    expect(result).toMatchSnapshot()
  })
})
