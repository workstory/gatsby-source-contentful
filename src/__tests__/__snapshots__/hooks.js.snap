// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useContentfulImage constrained: width and aspectRatio 1`] = `
Object {
  "backgroundColor": null,
  "height": 337,
  "images": Object {
    "fallback": Object {
      "sizes": "(min-width: 600px) 600px, 100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=337",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=150&h=84 150w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=300&h=169 300w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=337 600w",
    },
    "sources": Array [],
  },
  "layout": "constrained",
  "width": 600,
}
`;

exports[`useContentfulImage constrained: width and height 1`] = `
Object {
  "backgroundColor": null,
  "height": 480,
  "images": Object {
    "fallback": Object {
      "sizes": "(min-width: 600px) 600px, 100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=480",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=150&h=120 150w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=300&h=240 300w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=480 600w",
    },
    "sources": Array [],
  },
  "layout": "constrained",
  "width": 600,
}
`;

exports[`useContentfulImage fixed: width and height 1`] = `
Object {
  "backgroundColor": null,
  "height": 480,
  "images": Object {
    "fallback": Object {
      "sizes": "600px",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=480",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=480 600w",
    },
    "sources": Array [],
  },
  "layout": "fixed",
  "width": 600,
}
`;

exports[`useContentfulImage fullWidth: aspect ratio 1`] = `
Object {
  "backgroundColor": null,
  "height": 0.5625,
  "images": Object {
    "fallback": Object {
      "sizes": "100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=4096&h=2304",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=180 320w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=654&h=368 654w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=768&h=432 768w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1024&h=576 1024w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1366&h=768 1366w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1600&h=900 1600w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1920&h=1080 1920w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=2048&h=1152 2048w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=2560&h=1440 2560w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=3440&h=1935 3440w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=3840&h=2160 3840w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=4096&h=2304 4096w",
    },
    "sources": Array [],
  },
  "layout": "fullWidth",
  "width": 1,
}
`;

exports[`useContentfulImage fullWidth: aspectRatio, maxWidth 1`] = `
Object {
  "backgroundColor": null,
  "height": 0.5625,
  "images": Object {
    "fallback": Object {
      "sizes": "100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=4096&h=2304",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=180 320w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=654&h=368 654w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=768&h=432 768w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1024&h=576 1024w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1366&h=768 1366w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1600&h=900 1600w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1920&h=1080 1920w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=2048&h=1152 2048w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=2560&h=1440 2560w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=3440&h=1935 3440w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=3840&h=2160 3840w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=4096&h=2304 4096w",
    },
    "sources": Array [],
  },
  "layout": "fullWidth",
  "width": 1,
}
`;

exports[`useContentfulImage fullWidth: width, height & maxWidth 1`] = `
Object {
  "backgroundColor": null,
  "height": 0.75,
  "images": Object {
    "fallback": Object {
      "sizes": "100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=800&h=600",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=240 320w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=654&h=491 654w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=768&h=576 768w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=800&h=600 800w",
    },
    "sources": Array [],
  },
  "layout": "fullWidth",
  "width": 1,
}
`;
